crd-docs/bin/validate-kiali-cr.sh

```diff
@@ -0,0 +1,180 @@
+      cat <<HELPMSG
+
+$0 [option...]
+
+  -ce|--client-exe
+      The path to the client executable. Should be a path to either a "oc" or "kubectl" executable.
+  -crd|--crd-location
+      The file or URL location where the Kiali CRD is. This CRD must include the schema.
+      If not specified, the internally defined CRD is used.
+  -kcf|--kiali-cr-file
+      The file of the Kiali CR to test.
+  -kcn|--kiali-cr-name
+      The name of an existing Kiali CR to test.
+  -n|--namespace
+      The namespace where the existing CR is or where the test CR will be created.
+      Default: "default"
+  -pc|--print-crd
+      If true, then this script will just print the CRD used to validate. It will not validate anything.
+      Default "false"
+HELPMSG
+      exit 1
+      ;;
+    *)
+      echo "Unknown argument [$key]. Aborting."
+      exit 1
+      ;;
+  esac
+done
+
+# Set up some defaults
+
+: ${NAMESPACE:=default}
+: ${PRINT_CRD:=false}
+
+# If we are to print the CRD, do it now immediately and then exit. Nothing else to do.
+if [ "${PRINT_CRD}" == "true" ]; then
+  echo "$(crd)"
+  exit $?
+fi
+
+echo "=== SETTINGS ==="
+echo KIALI_CRD_LOCATION=${KIALI_CRD_LOCATION:-}
+echo KIALI_CR_FILE=${KIALI_CR_FILE:-}
+echo KIALI_CR_NAME=${KIALI_CR_NAME:-}
+echo NAMESPACE=${NAMESPACE}
+echo PRINT_CRD=${PRINT_CRD}
+echo "=== SETTINGS ==="
+
+# Determine what cluster client tool we are using.
+if [ -z "${CLIENT_EXE:-}" ]; then
+  if which oc &>/dev/null; then
+    CLIENT_EXE="$(which oc)"
+    echo "Using 'oc' located here: ${CLIENT_EXE}"
+  else
+    if which kubectl &>/dev/null; then
+      CLIENT_EXE="$(which kubectl)"
+      echo "Using 'kubectl' located here: ${CLIENT_EXE}"
+    else
+      echo "ERROR! You do not have 'oc' or 'kubectl' in your PATH. Please install it and retry."
+      exit 1
+    fi
+  fi
+else
+  echo "Client executable: ${CLIENT_EXE}"
+fi
+
+if [ -z "${KIALI_CR_FILE:-}" -a -z "${KIALI_CR_NAME:-}" ]; then
+  echo "ERROR! You must specify one of either --kiali-cr-file or --kiali-cr-name"
+  exit 1
+fi
+
+if [ -n "${KIALI_CR_FILE:-}" -a -n "${KIALI_CR_NAME:-}" ]; then
+  echo "ERROR! You must specify only one of either --kiali-cr-file or --kiali-cr-name"
+  exit 1
+fi
+
+if [ -n "${KIALI_CR_FILE:-}" -a ! -f "${KIALI_CR_FILE:-}" ]; then
+  echo "ERROR! Kiali CR file is not found: [${KIALI_CR_FILE:-}]"
+  exit 1
+fi
+
+if [ -n "${KIALI_CR_NAME:-}" ]; then
+  if ! ${CLIENT_EXE} get -n "${NAMESPACE}" kiali "${KIALI_CR_NAME}" &> /dev/null; then
+    echo "ERROR! Kiali CR [${KIALI_CR_NAME}] does not exist in namespace [${NAMESPACE}]"
+    exit 1
+  fi
+fi
+
+# Make sure we have admin rights to some cluster
+if ! ${CLIENT_EXE} get namespaces &> /dev/null ; then
+  echo "ERROR! You must be connected to/logged into a cluster"
+  exit 1
+fi
+if [ "$(${CLIENT_EXE} auth can-i create crd --all-namespaces)" != "yes" ]; then
+  echo "ERROR! You must have cluster-admin permissions"
+  exit 1
+fi
+
+# install the test CRD with the schema
```

Comment:
I wish there was a way to validate a CR against a CRD's openapi schema offline. I'm a little leery of creating/deleting a new CRD and CR to validate. If an error happens before cleanup you could end up with two kiali CRDs installed in your cluster and unknowingly use the testing one. There doesn't seem to be a way to do offline validation with the standard tooling though. And this is mostly a diagnostic tool anyways.

Reply:
I agree with all of the above. And I did think about the case where, for some reason, the test CRD fails to delete. It will be annoying because you will need to manually delete it. However, do note - the test CRD that is created by this tool is literally called "TestKiali" - it is NOT the same name as the CRD that the operator/helm will create and is NOT the CRD that will be used by the Kiali operator itself. So, yes, if the test CRD fails to delete there will be two CRDs, but the TestKiali CRD is essentially unused. The operator ignores it completely. You can create CRs of the kind "TestKiali" (which is what this tool does) and those objects do nothing - they will just float around without any impact to Kiali itself.